nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_prototype.rs

```diff
@@ -3084,14 +3084,107 @@ impl ArrayPrototype {
         call_function(agent, gc, func, array.into_value(), None)
     }
 
+    /// ### [23.1.3.37 Array.prototype.unshift ( ...items )](https://tc39.es/ecma262/#sec-array.prototype.unshift)
+    ///
+    /// This method prepends the arguments to the start of the array, such that
+    /// their order within the array is the same as the order in which they appear
+    /// in the argument list.
+    ///
+    /// > ### Note
+    /// >
+    /// > This method is intentionally generic; it does not require that its
+    /// > this value be an Array. Therefore it can be transferred to other
+    /// > kinds of objects for use as a method.
     fn unshift(
-        _agent: &mut Agent,
-        _gc: GcScope<'_, '_>,
-
-        _this_value: Value,
-        _: ArgumentsList,
+        agent: &mut Agent,
+        mut gc: GcScope<'_, '_>,
+        this_value: Value,
+        items: ArgumentsList,
     ) -> JsResult<Value> {
-        todo!();
+        // 1. Let O be ? ToObject(this value).
+        let o = to_object(agent, this_value)?;
+        // 2. Let len be ? LengthOfArrayLike(O).
+        let len = length_of_array_like(agent, gc.reborrow(), o)?;
+        // 3. Let argCount be the number of elements in items.
+        let arg_count = items.len();
+        // 4. If argCount > 0, then
+        if arg_count > 0 {
+            // a. If len + argCount > 2**53 - 1, throw a TypeError exception.
+            if (len + arg_count as i64) > SmallInteger::MAX_NUMBER {
+                return Err(agent.throw_exception_with_static_message(
+                    ExceptionType::TypeError,
+                    "Array length overflow",
+                ));
+            }
+            if let Value::Array(array) = this_value {
```

Comment:
**issue:** Sorry I didn't notice this at first, but the fast path should be at the very beginning of the function: Before even doing `o = to_object` etc. What we want to optimise for is something like `[b,c,d].unshift(a)` as that is the common case, and there the `to_object` call is just going to return `this_value` directly anyway so that call isn't useful.